### PR TITLE
Torch based DDPG

### PR DESCRIPTION
**Pull request by ShuaibinLi**
### Add files: 
mujoco_agent.py, mujoco_model.py, train.py, README.md, ---> benchmark/torch/
                 ddpg.py ---> parl/algorithms/torch/
### Change file: 
parl/algorithms/torch/__init__.py

### CODE REVIEW

**Comment by TomorrowIsAnOtherDay** [comment]
Why do you close and reopen a new PR again?

File: benchmark/torch/ddpg/train.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+import gym
+import argparse
+from parl.utils import logger, tensorboard, ReplayMemory
+from mujoco_model import MujocoModel
+from mujoco_agent import MujocoAgent
+from parl.algorithms import DDPG
+
+
+# Run episode for training
+def run_train_episode(agent, env, rpm, args):
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py]: args is a global var

File: benchmark/torch/ddpg/train.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        state, done = eval_env.reset(), False
+        while not done:
+            action = agent.predict(np.array(state))
+            state, reward, done, _ = eval_env.step(action)
+            avg_reward += reward
+    avg_reward /= eval_episodes
+    return avg_reward
+
+
+def main(args):
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py]: args is a global var

Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: Resolved.
image

File: benchmark/torch/ddpg/train.py (hunk 1 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: Resolved.
image

**Comment by ShuaibinLi** [comment]
> Why do you close and reopen a new PR again?
I've made too much commits and want to make it more clear.

**Comment by TomorrowIsAnOtherDay** [comment]
> > Why do you close and reopen a new PR again?
> > I've made too much commits and want to make it more clear.

Don't worry. Just left the comments and commit records in a PR such that we can see how you improve the code.

File: parl/algorithms/torch/ddpg.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        self.target_model = deepcopy(self.model)
+        self.actor_optimizer = torch.optim.Adam(
+            self.model.get_actor_params(), lr=actor_lr)
+        self.critic_optimizer = torch.optim.Adam(
+            self.model.get_critic_params(), lr=critic_lr)
+
+    def predict(self, obs):
+        return self.model.policy(obs)
+
+    def learn(self, obs, action, reward, next_obs, terminal):
```
Comment (rical730) [inline parl/algorithms/torch/ddpg.py:51]: please split `learn()` function to `_actor_learn()` and `_critic_learn()`, it will be more readable.

File: benchmark/torch/ddpg/train.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            tensorboard.add_scalar('eval/episode_reward', avg_reward,
+                                   total_steps)
+            logger.info('Evaluation over: {} episodes, Reward: {}'.format(
+                args.eval_episodes, avg_reward))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--env", default="HalfCheetah-v1")  # OpenAI gym environment name
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py]: please move comment into `help` argument

File: parl/algorithms/torch/ddpg.py (hunk 3 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/ddpg.py:51]: done

File: benchmark/torch/ddpg/train.py (hunk 4 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: done

File: parl/algorithms/torch/ddpg.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        self._critic_learn(obs, action, reward, next_obs, terminal)
+        self._actor_learn(obs)
+
+    def _critic_learn(self, obs, action, reward, next_obs, terminal):
+        self.total_it += 1
```
Comment (rical730) [inline parl/algorithms/torch/ddpg.py]: please handle `self.total_it` in learn()

File: parl/algorithms/torch/ddpg.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        critic_loss = F.mse_loss(current_Q, target_Q)
+
+        # Optimize the critic
+        self.critic_optimizer.zero_grad()
+        critic_loss.backward()
+        self.critic_optimizer.step()
+
+    def _actor_learn(self, obs):
+        # Update the frozen target models
+        if self.total_it % self.policy_freq == 0:
```
Comment (rical730) [inline parl/algorithms/torch/ddpg.py]: please handle `self.total_it` in learn()

Comment (ShuaibinLi) [reply parl/algorithms/torch/ddpg.py]: done

File: parl/algorithms/torch/ddpg.py (hunk 5 above)
Comment (ShuaibinLi) [reply parl/algorithms/torch/ddpg.py]: done

File: benchmark/torch/ddpg/train.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    parser.add_argument(
+        "--batch_size", default=100, type=int, help='Batch size for learning')
+    parser.add_argument("--discount", default=0.99, help='Discount factor')
+    parser.add_argument(
+        "--tau", default=0.005, help='Target network update rate')
+    parser.add_argument(
+        "--actor_lr", default=1e-3, help='Learning rate of actor network')
+    parser.add_argument(
+        "--critic_lr", default=1e-3, help='Learning rate of critic network')
+    parser.add_argument(
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: Remove this argument.

File: parl/algorithms/torch/ddpg.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(
+            self,
+            model,
+            max_action,
+            discount=None,
+            tau=None,
+            actor_lr=None,
+            critic_lr=None,
+            decay=None,
+            policy_freq=1,
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/torch/ddpg.py]: remove this argument.

File: benchmark/torch/ddpg/train.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        "--start_timesteps",
+        default=1e4,
+        type=int,
+        help='Time steps initial random policy is used')
+    parser.add_argument(
+        "--eval_freq",
+        default=5e3,
+        type=int,
+        help='How often (time steps) to evaluate')
+    parser.add_argument(
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: remove this argument.

File: benchmark/torch/ddpg/train.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        "--eval_episodes",
+        default=5,
+        type=int,
+        help='How many episodes during evaluation')
+    parser.add_argument(
+        "--max_timesteps",
+        default=1e6,
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: 1e6 -> 5e6

File: benchmark/torch/ddpg/train.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        discount=args.discount,
+        tau=args.tau,
+        actor_lr=args.actor_lr,
+        critic_lr=args.critic_lr,
+        policy_freq=args.policy_freq)
+    agent = MujocoAgent(algorithm, state_dim, action_dim)
+    rpm = ReplayMemory(
+        max_size=int(1e6), obs_dim=state_dim, act_dim=action_dim)
+
+    episode_num = 0
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: Remove this counter. We often use the environment step as the counter.

File: benchmark/torch/ddpg/mujoco_model.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+'''
+Model of DDPG:  defines an Actor/policy network given obs as input,
+                      & a Critic/value network given obs and action as input.
+'''
+
+
+class MujocoModel(parl.Model):
+    def __init__(self, obs_dim, action_dim, max_action):
+        super(MujocoModel, self).__init__()
+        self.actor_model = Actor(obs_dim, action_dim, max_action)
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/mujoco_model.py]: Do not pass max_action here. Mapping the action into the value range of the environment can be finished in the environment wrapper.
Please refer to the implementation in TD3.
https://github.com/PaddlePaddle/PARL/blob/8877d68f87407c438ba079cd8d56303dfae95521/examples/TD3/train.py#L23

File: benchmark/torch/ddpg/train.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+
+# Run episode for training
+def run_train_episode(agent, env, rpm):
+    action_dim = env.action_space.shape[0]
+    max_action = float(env.action_space.high[0])
+    obs, done = env.reset(), False
+    episode_reward, episode_steps = 0, 0
+
+    while not done:
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py:43]: The `done` can be true if the environment steps exceed maximum step:1000, while it does not represent the end of an episode.

File: benchmark/torch/ddpg/train.py (hunk 9 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: done

File: benchmark/torch/ddpg/train.py (hunk 10 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: done

File: benchmark/torch/ddpg/mujoco_model.py (hunk 12 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/mujoco_model.py]: done

File: benchmark/torch/ddpg/train.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    avg_reward /= eval_episodes
+    return avg_reward
+
+
+def main():
+    logger.info("------------------ DDPG ---------------------")
+    logger.info('Env: {}, Seed: {}'.format(args.env, args.seed))
+    logger.info("---------------------------------------------")
+
+    env = gym.make(args.env)
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py:91]: Why don't you use the same environment for training and evaluation ?

File: benchmark/torch/ddpg/train.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+TAU = 0.005
+ACTOR_LR = 1e-3
+CRITIC_LR = 1e-3
+EXPL_NOISE = 0.1  # Std of Gaussian exploration noise
+
+
+# Run episode for training
+def run_train_episode(agent, env, rpm):
+    action_dim = env.action_space.shape[0]
+    obs, done = env.reset(), False
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]:
```python
obs = env.reset()
done = False
```

File: benchmark/torch/ddpg/train.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+    return episode_reward, episode_steps
+
+
+# Runs policy for 5 episodes by default and returns average reward
+# A fixed seed is used for the eval environment
+def run_evaluate_episodes(agent, eval_env, eval_episodes):
+    avg_reward = 0.
+    for _ in range(eval_episodes):
+        state, done = eval_env.reset(), False
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: Do not combine them in a single line.

File: benchmark/torch/ddpg/train.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+from mujoco_model import MujocoModel
+from mujoco_agent import MujocoAgent
+from parl.algorithms import DDPG
+
+WARMUP_STEPS = 1e4
+EVAL_EVERY_STEPS = 5e3
+EVAL_EPISODES = 5
+MEMORY_SIZE = int(1e6)
+BATCH_SIZE = 100
+DISCOUNT = 0.99
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: rename discount to gamma.

File: benchmark/torch/ddpg/README.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+### Benchmark result
+
+<img src=".benchmark/DDPG_results.png" width = "800" height ="400" alt="DDPG_results"/>
+
+## How to use
+### Dependencies:
++ python3.5+
++ [parl](https://github.com/PaddlePaddle/PARL)
++ gym
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/README.md]: Please specify the version for the gym library using `==`.

File: benchmark/torch/ddpg/README.md [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
++ torch
++ mujoco-py>=1.50.1.0
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/README.md]: Please specify the version for the mujoco-py library using `==`.

File: benchmark/torch/ddpg/mujoco_agent.py [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+        self.device = torch.device("cuda" if torch.cuda.
+                                   is_available() else "cpu")
+
+        self.alg.sync_target(decay=0)
+
+    def predict(self, obs):
+        obs = torch.FloatTensor(obs.reshape(1, -1)).to(self.device)
+        action = self.alg.predict(obs)
+        action_numpy = action.cpu().data.numpy().flatten()
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/mujoco_agent.py]: Use `.detach()` to obtain data, rather than `.data`.

File: benchmark/torch/ddpg/train.py (hunk 17 above)
Comment (TomorrowIsAnOtherDay) [reply benchmark/torch/ddpg/train.py]: same in algorithms.ddpg

File: benchmark/torch/ddpg/mujoco_agent.py (hunk 20 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/mujoco_agent.py]: done

File: benchmark/torch/ddpg/README.md (hunk 19 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/README.md]: done

File: benchmark/torch/ddpg/mujoco_agent.py (hunk 20 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/mujoco_agent.py]: done

File: benchmark/torch/ddpg/train.py (hunk 15 above)
Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: done

File: benchmark/torch/ddpg/train.py [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+    done = False
+    episode_reward, episode_steps = 0, 0
+
+    while not done:
+        episode_steps += 1
+        # Select action randomly or according to policy
+        if rpm.size() < WARMUP_STEPS:
+            action = env.action_space.sample()
+        else:
+            predict_action = agent.predict(np.array(obs))
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: Use `agent.sample` for exploration.

Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: done

File: benchmark/torch/ddpg/train.py [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+    done = False
+    episode_reward, episode_steps = 0, 0
+
+    while not done:
+        episode_steps += 1
+        # Select action randomly or according to policy
+        if rpm.size() < WARMUP_STEPS:
+            action = np.random.uniform(-1, 1, size=action_dim)
+        else:
+            action = agent.sample(np.array(obs))
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: Obs has been a `np.array`.

File: benchmark/torch/ddpg/train.py [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+
+# Runs policy for 5 episodes by default and returns average reward
+# A fixed seed is used for the eval environment
+def run_evaluate_episodes(agent, env, eval_episodes):
+    avg_reward = 0.
+    for _ in range(eval_episodes):
+        obs = env.reset()
+        done = False
+        while not done:
+            action = agent.predict(np.array(obs))
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/torch/ddpg/train.py]: Obs has been a `np.array`.

Comment (ShuaibinLi) [reply benchmark/torch/ddpg/train.py]: done

**Review by rical730 — APPROVED** [review]
Looks good to me.
